### PR TITLE
Refactor cpp fixes in preparation for C++ autosummary extension

### PR DESCRIPTION
**Pull request by jbms**

### CODE REVIEW

File: .pylintrc [hunk 1]
```diff
@@ -3,7 +3,7 @@
 # A comma-separated list of package or module names from where C extensions may
 # be loaded. Extensions are loading into the active Python interpreter and may
 # run arbitrary code.
-extension-pkg-allow-list=
+extension-pkg-allow-list=pydantic.dataclasses
```
Comment (2bndy5) [inline .pylintrc:6]: I had to add `lxml` to this as well, but that depends on your C++ apigen implementation.

**Comment by 2bndy5** [comment]
I'm a bit confused why this is a separate PR since its geared directly for #117

**Comment by 2bndy5** [comment]
Kinda makes sense, but it seems like a complex rebasing tactic. Should this be made a draft and kept up in parallel to #117 (for as long as that's open)?

**Comment by jbms** [comment]
I plan to keep the `refactor-cpp-fixes` branch (this PR) as an ancestor of `cpp-apigen` (#117).

https://github.com/jbms/sphinx-immaterial/network

I'm hoping though that this can be merged relatively soon while we still revise #117, since I don't think further revisions of #117 will require additional changes here.

**Review by 2bndy5 — CHANGES_REQUESTED** [review]
Thanks for splitting this up into bite sized extensions (much easier to review this way).
### Possible separate issues

I wouldn't mind seeing a new `cpp:namespace-decl` directive to actually allow documenting a namespace declaration (even if not using the upcoming `cpp.apigen` ext). I think breathe uses a modified type declaration for this.

I also noticed that `cpp:function` doesn't support exception specifications like
```cpp
void function(int n) throws(std::out_of_range);
```
While I understand that this is generally discouraged due to differences in compiler behavior, it is still part of the C++ specs.

File: sphinx_immaterial/apidoc/cpp/parameter_objects.py [hunk 2]
```diff
@@ -0,0 +1,546 @@
...
+        if not unique_decls:
+            all_params = {}
+            for sig_param_nodes in sig_param_nodes_for_signature:
+                all_params.update(sig_param_nodes)
+            logger.warning(
+                "Parameter name %r does not match any of the parameters "
+                "defined in the signature: %r",
+                param_name,
+                list(all_params.keys()),
+                location=param_node,
```
[the remark this answers is not in the document]
Comment (jbms) [reply sphinx_immaterial/apidoc/cpp/parameter_objects.py]: Fixed.

File: sphinx_immaterial/apidoc/cpp/cpp_resolve_c_xrefs.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+POSSIBLE_MACRO_TARGET_PATTERN = re.compile("^[A-Z]+[A-Z_0-9]*(?:::[a-zA-Z0-9_]+)?$")
+"""Pattern for targets that may possibly refer to a macro or macro parameter.
+
+Any targets matching this pattern that are not defined as C++ symbols will be
+looked up in the C domain.
+
+For efficiency, this is restricted to names that consist of only uppercase
+letters and digits, to avoid a duplicate C domain query for all symbols that
+will be resolved by `sphinx_immaterial.external_cpp_references`.
```
[the remark this answers is not in the document]
Comment (jbms) [reply sphinx_immaterial/apidoc/cpp/cpp_resolve_c_xrefs.py]: I changed this to a config option and documented it, so now users can change the pattern if they want.

**Comment by jbms** [comment]
> Thanks for splitting this up into bite sized extensions (much easier to review this way).
> 
> ### Possible separate issues
> I wouldn't mind seeing a new `cpp:namespace-decl` directive to actually allow documenting a namespace declaration (even if not using the upcoming `cpp.apigen` ext). I think breathe uses a modified type declaration for this.

Yes that could be added to the C++ domain.

Separate from that, it would be nice to modify the C++ domain to suppress the nitpick reference warnings for every namespace, though.

However, I think that might be kind of annoying to monkey patch in --- maybe it can be, though.

> 
> I also noticed that the `cpp:function` directive doesn't support exception specifications like
> 
> ```c++
> void function(int n) throws(std::out_of_range);
> ```
> 
> While I understand that this is generally discouraged due to differences in compiler behavior, it is still part of the C++ specs.

I imagine this could be supported by the C++ domain easily enough.  But note that it was deprecated in C++11 and removed in C++17:

https://en.cppreference.com/w/cpp/language/except_spec

File: sphinx_immaterial/apidoc/cpp/cpp_resolve_c_xrefs.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (2bndy5) [reply sphinx_immaterial/apidoc/cpp/cpp_resolve_c_xrefs.py]: oh awesome!

**Comment by 2bndy5** [comment]
> I imagine this could be supported by the C++ domain easily enough. But note that it was deprecated in C++11 and removed in C++17

This is why I haven't taken the issue upstream. Still the spec compatibility can be set at build time (like tensorstore does).

**Review by 2bndy5 — APPROVED** [review]

**Pull request merged**